scene with platforms

sander76 opened the pull request:
Implements scene as a python package. Allowing for other scenes (from other systems) to be included as a platform. It defaults to a home assistant scene.

sander76 committed:
first commit

sander76 commented:
How can I remove the scene test fail as you said in the other pull request ? (which I closed.)

sander76 committed:
flake8 correction

balloob commented on homeassistant/components/scene/homeassistant.py:
```diff
@@ -92,37 +70,23 @@ def _process_config(scene_config):
         self.hass = hass
         self.scene_config = scene_config
-
         self.update()
```
No need to call update

balloob commented:
Can you make sure commit messages make more sense?

Tests can be found in the tests directory

sander76 committed:
- removed update method
- removed failing tests from test_scene

sander76 commented:
I will. Sorry about that.

balloob commented on homeassistant/components/scene/__init__.py:
```diff
@@ -0,0 +1,85 @@
+        return False
+
+    @property
+    def name(self):
```
No need for this.

balloob commented on homeassistant/components/scene/__init__.py:
```diff
@@ -0,0 +1,85 @@
+        return STATE
+
+    @property
+    def entity_ids(self):
```
No need for this.

sander76 replied:
@balloob you mean ```entity_ids``` and the ```name``` property?

balloob replied:
Yes, they are not needed for the generic entity. `name` is already defined in `Entity` and `entity_ids` is an implementation detail of the home assistant platform.

sander76 committed:
removed unnecessary properties.

balloob commented on homeassistant/components/scene/__init__.py:
```diff
@@ -0,0 +1,76 @@
+    for entry in config:
+        if DOMAIN in entry:
+            if not any(CONF_PLATFORM in key for key in config[entry]):
+                config[entry] = {'platform': 'homeassistant',
```
Did you test this code locally? I am pretty sure that this will fail because the config that is passed in when loading HA normally is immutable.

balloob commented on homeassistant/components/scene/__init__.py:
```diff
@@ -0,0 +1,76 @@
+
+    logger = logging.getLogger(__name__)
+
+    for entry in config:
```
Why are you looping through the entries? The only part of the config that you should care about is `config[DOMAIN]`. The if-statement on the next line also doesn't make sense as we do not care about the key.

```python
if not any(CONF_PLATFORM in key for key in config.get(DOMAIN, {}):
    config = {DOMAIN: {'platform': 'homeassistant', config[DOMAIN]}
```

sander76 replied:
Considering the following configuration file:

```
scene test:
  - name: All on
    entities:
      input_boolean.notify_home: on

scene:
  - name: All off
    entities:
      input_boolean.notify_home: off
```

doing ```config.get(DOMAIN,{})``` will get me only ```scene``` and not ```scene test```.
So I thought looping through all config entries will make sure I get hold of all ```scene``` domains and make sure they get a platform attached.

I did test it locally and it works. (I also ran it through a debugger and the config does get the changes even though it is to be immutable.)

balloob replied:
That feature was undocumented so I feel comfortable not making it backwards
compatible.

Paulus

On Tue, Feb 23, 2016 at 11:47 PM, sander76 <notifications@github.com> wrote:

> In homeassistant/components/scene/__init__.py
> <https://github.com/balloob/home-assistant/pull/1374#discussion_r53903819>
> :
>
> > +def activate(hass, entity_id=None):
> > +    """ Activate a scene. """
> > +    data = {}
> > +
> > +    if entity_id:
> > +        data[ATTR_ENTITY_ID] = entity_id
> > +
> > +    hass.services.call(DOMAIN, SERVICE_TURN_ON, data)
> > +
> > +
> > +def setup(hass, config):
> > +    """ Sets up scenes. """
> > +
> > +    logger = logging.getLogger(__name__)
> > +
> > +    for entry in config:
>
> Considering the following configuration file:
>
> scene test:
>   - name: All on
>     entities:
>       input_boolean.notify_home: on
>
> scene:
>   - name: All off
>     entities:
>       input_boolean.notify_home: off
>
> doing config.get(DOMAIN,{}) will get me only scene and not scene test.
> So I thought looping through all config entries will make sure I get hold
> of all scene domains and make sure they get a platform attached.
>
> I did test it locally and it works. (I also ran it through a debugger and
> the config does get the changes even though it is to be immutable.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1374/files#r53903819>.
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.

sander76 replied:
Look here:

https://home-assistant.io/getting-started/devices/#style-2

sander76 replied:
@balloob Hi Paulus, looking at my last remark, what do you prefer to do?

balloob replied:
I see, I guess you're right. We have a helper method that will extract the keys for you: [`extract_domain_configs`](https://github.com/balloob/home-assistant/blob/dev/homeassistant/helpers/__init__.py#L61-L64).

The value of a config can besides being a list also be a dict. If that is the case just wrap the value in a list.

_Disclaimer: I did not test this code_
```python
    # You are not allowed to mutate the original config so make a copy
    config = dict(config)

    for config_key in extract_domain_configs(config, domain):
        platform_config = config[config_key]
        if not isinstance(platform_config, list):
            platform_config = [platform_config]

        if not any(CONF_PLATFORM in entry for entry in platform_config):
            platform_config = [{'platform': 'homeassistant', 'config': entry}
                               for entry in platform_config]

        config[config_key] = platform_config
```

sander76 commented:
@balloob 
Your code makes a config like this:
```
# this defaults to homeassistant platform.
scene:
  - name: All on
    platform: homeassistant
    entities:
      input_boolean.notify_home: on

  - name: All off
    platform: homeassistant
    entities:
      input_boolean.notify_home: off
```
That shouldn't be ? There should be one unique platform. Not two. Also your codes requires a rewrite of the actual homeassistant scene platform as it expects a list of scene entities.

In my original approach I would allow only one platform in a domain. Another platform would require another. Like the ```scene``` and ```scene 1``` in a config file.

What do you want me to do ? Rewrite the homeassistant scene platform or rewrite your code ?

balloob commented:
There is no requirement to have to load each platform once. It's totally fine to load a platform multiple times.

`EntityComponent.setup` calls `config_per_platform` which will convert [all dictionaries to lists](https://github.com/balloob/home-assistant/blob/dev/homeassistant/helpers/__init__.py#L45-L46).

sander76 committed:
1. added platform per scene entry.
2. changed homeassistant scene setup_platform method to work with 1.

sander76 committed:
removed logger.
I guess I should add some error checking, but I'd like you to okay this code first.

sander76 committed:
removed logger.
I guess I should add some error checking, but I'd like you to okay this code first.

balloob commented on homeassistant/components/scene/homeassistant.py:
```diff
@@ -25,41 +22,12 @@
-            scene.activate()
-
-    hass.services.register(DOMAIN, SERVICE_TURN_ON, handle_scene_service)
+    add_devices([HomeAssistantScene(hass, _process_config(scene_config))])
```
People might still want to put in lists. So let's do this:

```yaml
if not isinstance(scene_config, list):
    scene_config = [scene_config]

add_devices(HomeAssistantScene(hass, _process_config(scene))
            for scene in scene_config)
```

balloob commented on homeassistant/components/scene/homeassistant.py:
```diff
@@ -25,41 +22,12 @@
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Sets up home assistant scene entries. """
+    scene_config = config.get("config")
```
First, I know I'm not consistent and are going against what I said earlier, sorry about that. But I think that `config` is not a good description. Let's call it `states`.

sander76 committed:
Added your suggestions.
Looking at your code suggestion below I am not sure exactly how other people might want to put in lists. (But I am missing a more general overview of the code)

 ```
if not isinstance(scene_config,list):
  scene_config=[scene_config]
```

But it is there !
And changed "config" to "states" !

balloob commented:
Looks good :dolphin:  ! :tada: :tada:
